src-electron/db/query-upgrade.js

```diff
@@ -0,0 +1,126 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+const dbApi = require('./db-api.js')
+const dbEnum = require('../../src-shared/db-enum.js')
+const fs = require('fs')
+const fsp = fs.promises
+
+async function selectClusterName(db, clusterRef) {
+  let clusterName = await dbApi.dbAll(
+    db,
+    'SELECT NAME FROM CLUSTER WHERE CLUSTER_ID = ?',
+    [clusterRef]
+  )
+  return clusterName[0].NAME
+}
+
+async function selectAttributeName(db, attributeId) {
+  let attributeName = await dbApi.dbAll(
+    db,
+    'SELECT NAME FROM ATTRIBUTE WHERE ATTRIBUTE_ID = ?',
+    [attributeId]
+  )
+  return attributeName[0].NAME
+}
+
+async function queryMetaFile(db, packageId) {
+  let path = await dbApi.dbAll(
+    db,
+    'SELECT PATH FROM PACKAGE WHERE PACKAGE_ID = ?',
+    [packageId]
+  )
+  return path[0].PATH
+}
+async function queryPackages(db, attributeId) {
+  let package_ref = await dbApi.dbAll(
+    db,
+    'SELECT PACKAGE_REF FROM ATTRIBUTE WHERE ATTRIBUTE_ID = ?',
+    [attributeId]
+  )
+  return package_ref[0].PACKAGE_REF
+}
+
+/**
+ * Returns an array of objects containing global attributes that should be forced external.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} attributeId
+ * @returns An array of objects
+ */
+
+async function checkGlobals(db, attributeId) {
+  let pkgs = await queryPackages(db, attributeId)
+  let zcl = await queryMetaFile(db, pkgs)
+  let obj = await fsp.readFile(zcl)
+  let data = JSON.parse(obj)
+  let externals = data.attributeAccessInterfaceAttributes
+  let lists = data.listsUseAttributeAccessInterface
+  let forcedExternal = { externals, lists }
+  return forcedExternal
+}
+
+/**
+ * Returns a flag stating which type of storage option the attribute is categorized to be.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} staticAttribute
+ * @param {*} forcedExternal
+ * @param {*} clusterRef
+ * @param {*} attributeId
+ * @returns A flag.
+ */
+
+async function checkStorage(
+  db,
+  clusterName,
+  clusterRef,
+  storagePolicy,
+  forcedExternal,
+  attributeId,
+  attribute
+) {
+  let storageOption
+  let attributeName
+  if (!clusterName) {
+    clusterName = await selectClusterName(db, clusterRef)
+    if (storagePolicy == dbEnum.storagePolicy.attributeAccessInterface) {
+      storageOption = dbEnum.storageOption.external
+    } else if (storagePolicy == dbEnum.storagePolicy.any) {
+      storageOption = dbEnum.storageOption.ram
+    }
+  }
+  if (attributeId) {
+    attributeName = await selectAttributeName(db, attributeId)
+    if (forcedExternal.lists == true && attribute.type == 'array') {
+      storageOption = dbEnum.storageOption.external
+    }
+    if (
```

Comment:
@bzbarsky-apple this code which presumably changes global attributes to external if they are in the list seems to cause CI and codeine failures. Any idea what the failures mean?

Reply:
That code changes _all_ attributes to be external if they are of array type.  And that code is very much pre-existing, the right behavior, and not the problem here.

Reply:
okay thank you